File: src/vehicle.py
from enum import Enum
from typing import Optional

from motors import Motor, MotorDirection


class SpinDirection(Enum):
    """
    Defines different directions a vehicle can spin
    """
    LEFT = 3
    RIGHT = 4


class VehicleDirection(Enum):
    """
    Defines directions a vehicle can take. All four directions.
    Values are tuples, where the left item is the MotorDirection for
    the left Motor and the right item is the MotorDirection for the right Motor
    """
    FORWARD = MotorDirection.FORWARD
    BACKWARD = MotorDirection.BACKWARD
    LEFT = SpinDirection.LEFT
    RIGHT = SpinDirection.RIGHT

# Assert that all variants in VehicleDirection are unique
assert len(set(VehicleDirection.__members__.values())) == 4


class Vehicle:
    """ This class represents a Vehicle. In this case with 2 motors. """

    def __init__(self, left: Motor, right: Motor, default_speed: float):
        self.left = left
        self.right = right
        self.speed = 0.0
        self.default_speed = default_speed

    def move(
        self,
        direction: VehicleDirection,
        speed: Optional[float] = None
    ) -> float:
        """
        Move the vehicle in a direction, allows overwriting current speed

        direction: VehicleDirection in which to steer the vehicle to
        speed: Allow overwriting the current default speed for this move call

        return: Returns the speed of the vehicle before this call
        """
        previous_speed = self.speed
        self.speed = self.default_speed if speed is None else speed

        match direction:
            case VehicleDirection.FORWARD:
                self.left.move(MotorDirection.FORWARD, self.speed)
                self.right.move(MotorDirection.FORWARD, self.speed)
            case VehicleDirection.BACKWARD:
                self.left.move(MotorDirection.BACKWARD, self.speed)
                self.right.move(MotorDirection.BACKWARD, self.speed)
            case VehicleDirection.LEFT:
                self.left.move(MotorDirection.FORWARD, self.speed / 3)
                self.right.move(MotorDirection.FORWARD, self.speed)
            case VehicleDirection.RIGHT:
                self.left.move(MotorDirection.FORWARD, self.speed)
                self.right.move(MotorDirection.FORWARD, self.speed / 3)
            case _:
                raise NotImplementedError(
                    f"VehicleDirection not covered: {repr(direction)}"
                )

        return previous_speed


    def spin(
        self,
        direction: SpinDirection,
        speed: Optional[float] = None
    ) -> float:
        """
        Spin in a given direction

        direction: SpinDirection in which to spin to
        speed: Optional speed to overwrite the default speed

        return: Returns the speed of the vehicle before this call
        """
        previous_speed = self.speed
        self.speed = self.default_speed if speed is None else speed

        match direction:
            case SpinDirection.LEFT:
                self.left.move(MotorDirection.BACKWARD, self.speed)
                self.right.move(MotorDirection.FORWARD, self.speed)
            case SpinDirection.RIGHT:
                self.left.move(MotorDirection.FORWARD, self.speed)
                self.right.move(MotorDirection.BACKWARD, self.speed)
            case _:
                raise NotImplementedError(
                    f"SpinDirection not covered: {repr(direction)}"
                )

        return previous_speed

    def stop(self) -> float:
        """
        Stop the vehicle

        return: Returns the speed of the vehicle before the stop
        """
        previous_speed = self.speed
        self.speed = 0
        self.left.stop()
        self.right.stop()
        return previous_speed

File: crates/directions/src/vehicle.rs
use std::ops::Mul;

use crate::{MotorDirection, SpeedControl, SpinDirection, Stop};
use speed::Speed;

/// Represents directions a vehicle can take
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct VehicleDirection {
    /// The [`MotorDirection`] for the left motor
    pub left: MotorDirection,
    /// The [`MotorDirection`] for the right motor
    pub right: MotorDirection,
}

impl VehicleDirection {
    /// Create a new [`VehicleDirection`] from the individual [`MotorDirection`]s
    pub fn new(left: MotorDirection, right: MotorDirection) -> Self {
        Self { left, right }
    }
}

// Implement basic directions with a given [`Speed`]
impl VehicleDirection {
    /// [Forward](MotorDirection::Forward) direction with a given [`Speed`]
    pub fn forward(speed: Speed) -> Self {
        Self::new(
            MotorDirection::Forward(speed),
            MotorDirection::Forward(speed),
        )
    }

    /// [Backward](MotorDirection::Backward) direction with a given [`Speed`]
    pub fn backward(speed: Speed) -> Self {
        Self::new(
            MotorDirection::Backward(speed),
            MotorDirection::Backward(speed),
        )
    }

    /// Turn into a [`SpinDirection`] with a given [`Speed`]
    ///
    /// The [`Speed`] component of the [`SpinDirection`] is subtracted from the
    /// [`MotorDirection`] on the side on which the [`SpinDirection`] instructs
    pub fn turn(speed: Speed, direction: SpinDirection) -> Self {
        match direction {
            SpinDirection::Left(ratio) => Self::new(
                MotorDirection::Forward(speed).wrapping_sub_f64(2.0 * ratio.value()),
                MotorDirection::Forward(speed),
            ),
            SpinDirection::Right(ratio) => Self::new(
                MotorDirection::Forward(speed),
                MotorDirection::Forward(speed).wrapping_sub_f64(2.0 * ratio.value()),
            ),
        }
    }

    /// Spin the vehicle to the left in-place with a given [`Speed`]
    pub fn spin_left(speed: Speed) -> Self {
        Self::new(
            MotorDirection::Backward(speed),
            MotorDirection::Forward(speed),
        )
    }

    /// Spin the vehicle to the right in-place with a given [`Speed`]
    pub fn spin_right(speed: Speed) -> Self {
        Self::new(
            MotorDirection::Forward(speed),
            MotorDirection::Backward(speed),
        )
    }
}

impl Stop for VehicleDirection {
    fn is_stop(&self) -> bool {
        self.left.speed().value() == 0.0 && self.right.speed().value() == 0.0
    }
}

impl Mul<Speed> for VehicleDirection {
    type Output = Self;

    fn mul(self, rhs: Speed) -> Self::Output {
        Self::new(self.left * rhs, self.right * rhs)
    }
}

impl From<SpinDirection> for VehicleDirection {
    fn from(value: SpinDirection) -> Self {
        match value {
            SpinDirection::Left(speed) => VehicleDirection::new(
                MotorDirection::Backward(speed),
                MotorDirection::Forward(speed),
            ),
            SpinDirection::Right(speed) => VehicleDirection::new(
                MotorDirection::Forward(speed),
                MotorDirection::Backward(speed),
            ),
        }
    }
}
